File: Semester3/Advanced Methods of Programming/Toy Language Interpreter C#/Assignement7/Assignement7/Model/ReadStatement.cs
using System;
using System.Collections.Generic;
using System.IO;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace Assignement7.Model
{
    public class ReadStatement : IStatement
    {
        private Expression var;
        private String var_name;

        public ReadStatement(Expression var, String var_name)
        {
            this.var = var;
            this.var_name = var_name;
        }

        public override string ToString()
        {
            return "readFile(" + var.ToString() + ", " + var_name + ")";
        }


        public ProgramState Execute(ProgramState state)
        {
            int id = this.var.Evaluate(state.getSymbolTable());
            Tuple<string, TextReader> pair = state.getFileTable().Get(id);

            if(pair == null)
                throw new Exception("FileNotOpenedException at: " + this.ToString() + "\nNo such file descriptor: " + id.ToString());

            string line = pair.Item2.ReadLine();
            int val = 0;
            if (line != null)
                val = int.Parse(line);
            state.getSymbolTable().Put(this.var_name, val);
            return state;
        }
    }
}

File: Semester3/Advanced Methods of Programming/ExamenMAP/src/model/statements/file/ReadStatement.java
package model.statements.file;

import model.ProgramState;
import model.exceptions.FileException;
import model.expressions.Expression;
import model.statements.IStatement;
import model.util.MyFilePair;
import myCollections.MyIDictionary;

import java.io.BufferedReader;
import java.io.IOException;

public class ReadStatement implements IStatement {
    private Expression var;
    private String var_name;

    public ReadStatement(Expression var, String var_name) {
        this.var = var;
        this.var_name = var_name;
    }

    @Override
    public String toString() {
        return "readFile("+var.toString() +", "+ var_name +")";
    }

    @Override
    public ProgramState execute(ProgramState state) throws Exception {
        MyIDictionary<String, Integer> symbolTable = state.getSymbolTable();
        MyIDictionary<Integer, Integer> heapTable = state.getHeapTable();
        MyIDictionary<Integer, MyFilePair> fileTable = state.getFileTable();
        Integer uniq = var.evaluate(symbolTable, heapTable);

        if(!fileTable.containsKey(uniq))
            throw new FileException("there is no entry for this key!!!!\n");

        MyFilePair pair = fileTable.get(uniq);
        BufferedReader b = pair.getBufferedReader();
        String line = null;

        try{
            line = b.readLine();
        } catch (IOException e) {
            throw new FileException("cannot read from this file!!\n");
        }

        int value;
        if(line == null)
            value = 0;
        else
            try{
                value = Integer.parseInt(line);
            } catch (NumberFormatException e){
                value = 0;
            }
        symbolTable.put(var_name, value);

        return null;
    }
}
